File: src/neml2/models/solid_mechanics/crystal_plasticity/PlasticSpatialVelocityGradient.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "neml2/models/solid_mechanics/crystal_plasticity/PlasticSpatialVelocityGradient.h"
#include "neml2/models/crystallography/CrystalGeometry.h"

#include "neml2/tensors/Scalar.h"
#include "neml2/tensors/R2.h"
#include "neml2/tensors/SR2.h"
#include "neml2/tensors/SFFR4.h"
#include "neml2/tensors/list_tensors.h"
#include "neml2/tensors/functions/sum.h"

namespace neml2
{
register_NEML2_object(PlasticSpatialVelocityGradient);

OptionSet
PlasticSpatialVelocityGradient::expected_options()
{
  OptionSet options = Model::expected_options();

  options.doc() =
      "Caclulates the plastic spatial velocity gradient as \\f$ l^p = \\sum_{i=1}^{n_{slip}} "
      "\\dot{\\gamma}_i Q \\left(d_i \\otimes n_i \\right) Q^T "
      "\\f$ with \\f$ l^p \\f$ the plastic spatial velocity gradient, \\f$ \\dot{\\gamma}_i "
      "\\f$ the slip rate on the ith slip system, \\f$Q \\f$ the orientation, \\f$ d_i "
      "\\f$ the slip system direction, and \\f$ n_i \\f$ the slip system normal.";

  options.set_output("plastic_spatial_velocity_gradient") =
      VariableName(STATE, "internal", "plastic_spatial_velocity_gradient");
  options.set("plastic_spatial_velocity_gradient").doc() =
      "The name of the plastic spatial velocity gradient";

  options.set_input("orientation") = VariableName(STATE, "orientation_matrix");
  options.set("orientation").doc() = "The name of the orientation matrix tensor";

  options.set_input("slip_rates") = VariableName(STATE, "internal", "slip_rates");
  options.set("slip_rates").doc() = "The name of the tensor containg the current slip rates";

  options.set<std::string>("crystal_geometry_name") = "crystal_geometry";
  options.set("crystal_geometry_name").doc() =
      "The name of the Data object containing the crystallographic information for the material";

  return options;
}

PlasticSpatialVelocityGradient::PlasticSpatialVelocityGradient(const OptionSet & options)
  : Model(options),
    _crystal_geometry(register_data<crystallography::CrystalGeometry>(
        options.get<std::string>("crystal_geometry_name"))),
    _lp(declare_output_variable<R2>("plastic_spatial_velocity_gradient")),
    _R(declare_input_variable<R2>("orientation")),
    _g(declare_input_variable<Scalar>("slip_rates", _crystal_geometry.nslip()))
{
}

void
PlasticSpatialVelocityGradient::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  const auto lp_crystal = batch_sum(_g * _crystal_geometry.A(), -1);

  if (out)
    _lp = lp_crystal.rotate(_R);

  if (dout_din)
  {
    if (_g.is_dependent())
    {
      const auto dlp_dg = _crystal_geometry.A().rotate(R2(_R).batch_unsqueeze(-1));
      _lp.d(_g) = Tensor(dlp_dg.movedim(-3, -1), dlp_dg.batch_sizes().slice(0, -1));
    }

    if (_R.is_dependent())
      _lp.d(_R) = lp_crystal.drotate(_R);
  }
}
} // namespace neml2

File: neml2/models/solid_mechanics/crystal_plasticity/PlasticSpatialVelocityGradient.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Python-native mirror of the C++ ``PlasticSpatialVelocityGradient`` crystal-plasticity leaf."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from ....chain_rule import ChainRuleDict
from ....factory import register_native
from ....model import Model
from ....schema import HitSchema, dependency, input, output
from ....types import R2, Scalar, jvp_rotate, rotate, sum

if TYPE_CHECKING:
    from ....data import CrystalGeometry


@register_native("PlasticSpatialVelocityGradient")
class PlasticSpatialVelocityGradient(Model):
    r"""Caclulates the plastic spatial velocity gradient as
    $l^p = \sum_{i=1}^{n_{slip}} \dot{\gamma}_i Q \left(d_i \otimes n_i \right) Q^T$
    with $l^p$ the plastic spatial velocity gradient, $\dot{\gamma}_i$
    the slip rate on the ith slip system, $Q$ the orientation, $d_i$
    the slip system direction, and $n_i$ the slip system normal.
    """

    hit = HitSchema(
        input("orientation_matrix", R2, "The name of the orientation matrix"),
        input("slip_rates", Scalar, "The name of the tensor containg the current slip rates"),
        output(
            "plastic_spatial_velocity_gradient",
            R2,
            "The name of the plastic spatial velocity gradient",
        ),
        dependency(
            "crystal_geometry",
            "get_data",
            "The name of the Data object containing the crystallographic information "
            "for the material",
            default="crystal_geometry",
        ),
    )
    list_deriv = {("plastic_spatial_velocity_gradient", "slip_rates"): "dense"}

    def __init__(self, *, crystal_geometry: CrystalGeometry) -> None:
        super().__init__()
        self._cg = crystal_geometry

    def forward(  # type: ignore[override]
        self,
        R: R2,
        g: Scalar,
        v: ChainRuleDict | None = None,
    ):
        A = self._cg.A  # R2 sub_batch_ndim=1, the full d_i (x) n_i Schmid tensor
        weighted = A * g
        lp_cry = cast(R2, sum(weighted.sub_batch, -1))
        # Rotate to lab frame: l^p = R · lp_cry · R^T (no sym/skew projection).
        lp = rotate(lp_cry, R)
        if v is None:
            return lp

        # Differential pushforwards, mirror of PlasticDeformationRate:
        #   R:  d(lp) = jvp_rotate(lp_cry, R, dR)  (product rule, typed)
        #   gamma_dot: rotate(., R) is LINEAR in lp_cry = Sum_k gamma_dot_k A_k,
        #             so push the slip-rate tangent through the same sum + rotation,
        #             no Jacobian.
        def R_action(V: R2) -> R2:
            return jvp_rotate(lp_cry, R, V)

        def g_action(V: Scalar) -> R2:
            return rotate(cast(R2, sum((A * V).sub_batch, -1)), R)

        return lp, self.apply_chain_rule(
            v,
            "plastic_spatial_velocity_gradient",
            {"orientation_matrix": R_action, "slip_rates": g_action},
            output=lp,
        )
